### PR TITLE
Fix race condition on session Close()

### PR DESCRIPTION
**Pull request by SaveTheRbtz**
This is a proposed fix for libp2p/go-yamux#95.

Instead of accessing the `memory` field of all streams possibly racing with any other goroutines, I would propose wrapping MemoryManager with a goroutine-safe primitive that can track all of the allocated memory and release it on `Close()`.

I've used a lock instead of atomic on `allocated` to prevent races between multiple goroutines trying to allocate/release memory in parallel with `Close()` happening.

cc: @marten-seemann

### CODE REVIEW

**Review by marten-seemann — CHANGES_REQUESTED** [review]
What’s the advantage of doing it this way? The memory tracking has proved to be an incredibly brittle piece of code, and we finally managed to get it working. I’m hesitant to make any non-trivial changes to that code path.

**Comment by SaveTheRbtz** [comment]
This was the only way I could think of that would guarantee that there are no concurrent calls to the `MemoryManager`. Other solutions that I thought of did have race conditions between `Session.Close()` and all the background processes that can happen on streams in other goroutines.

**Comment by marten-seemann** [comment]
What about locking the mutex before calling `getWindowUpdate`? That seems to be the root of the race condition, no?

**Comment by SaveTheRbtz** [comment]
If I try to grab stream lock every consistently while calling `getWindowUpdate` I will introduce even more lock-order-reversals...  The current state of deadlocks is not great but it'll make it even worse.

Current state of deadlocks:
```
find . -name "*.go" | xargs gsed -i 's/\bsync.RWMutex\b/deadlock.RWMutex/'
find . -name "*.go" | xargs gsed -i 's/\bsync.Mutex\b/deadlock.Mutex/'
goimports -w .
```

**Comment by SaveTheRbtz** [comment]
I have branch with sendWindowUpdate split into get/send[1] w/o introducing additional potential deadlock but it has a couple of issues:
* `s.sendFlags()` can't be protected by a streamLock since it would introduce lock order reversal.  This means though that header construction is not atomic.
* `memoryManager` is still not fully synchronized with Session's `Close()` which means other threads can be inside it `Reserving`/`Releasing` memory while we are summing up memory to release inside the `Close()` function.

Basically, it looks like it would just hide the data race from race detector making it harder to spot.

[1] https://github.com/libp2p/go-yamux/compare/master...SaveTheRbtz:go-yamux:rbtz/lock?expand=1
